### PR TITLE
Add 2mb fork expiration date, version bit test cases, and bugfix

### PR DESCRIPTION
**Pull request by jtoomim**
- Adds an expiration date to the 2MB fork: Jan 1, 2018. After this date, no more votes will be cast for the fork. Blocks with FORK_BIT_2MB set (e.g. due to a different fork that recycles the bit) and with nTime after the expiration date will not trigger the fork.
- New test cases to ensure that the FORK_BIT_2MB and -vote2mb settings work properly.
- Fixes bug in ForkBits that prevented it from ever voting for the fork. (sizeForkTime.load() < nTime) should have been (nTime < sizeForkTime.load()).

### CODE REVIEW

File: qa/rpc-tests/bigblocks.py [hunk 1]
```diff
@@ -151,12 +155,17 @@ def copy_mempool(self, from_node, to_node):
...
         b1 = self.nodes[0].getblock(b1hash, True)
         assert(self.sync_blocks(self.nodes))
 
+        if expect_version:
+            assert b1['version'] & FORK_BLOCK_BIT
+        elif not expect_version==None:
+            assert not b1['version'] & FORK_BLOCK_BIT
+
         if expect_big:
             assert(b1['size'] > 1000*1000)
```
Comment (sugruedes) [inline qa/rpc-tests/bigblocks.py:170]: question: should 'expect_big' fail if less than 1mb?  im probably missing something, but it seems it should allow > and < 1mb

Comment (jtoomim) [inline qa/rpc-tests/bigblocks.py:170]: If you have 4 MB of transactions in your mempool (as is the case in bigblocks.py), and if your node is configured to make blocks up to 2 MB in size, and it does not generate a > 1 MB block when it is allowed to, then there is a bug and the test should fail.

Note that bigblocks.py is part of the unit test code. Its purpose is to help us find bugs. It should fail when anything happens other than what we want or expect to happen.

File: src/main.cpp [hunk 2]
```diff
@@ -5310,12 +5312,16 @@ uint32_t MaxLegacySigops(uint32_t nBlockTime)
...
 
-uint32_t ForkBits(uint32_t nTime) {
+uint32_t ForkBits(uint32_t nTime, const Consensus::Params& consensusParams) {
     uint32_t bits = 0;
     AssertLockHeld(cs_main);
     // Vote for 2 MB until the fork time to show continued support
-    if (sizeForkTime.load() < nTime && (GetBoolArg("-vote2mb", DEFAULT_2MB_VOTE) || (pblocktree->ForkBitActivated(FORK_BIT_2MB) != uint256())))
+    if (nTime < sizeForkTime.load() &&
+        nTime < consensusParams.SizeForkExpiration() &&
+        (GetBoolArg("-vote2mb", DEFAULT_2MB_VOTE) || (pblocktree->ForkBitActivated(FORK_BIT_2MB) != uint256())))
```
Comment (gavinandresen) [inline src/main.cpp:5321]: It would be nice to move this from main.cpp/main.h into miner.cpp, and not require the cs_main lock.

Can this be simplified to just:

If -vote2MB and time is before fork expiration time:  Vote for 2MB.

Any harm besides reserving the 0x10000000 bit until 2018 in doing that? I believe that bit is unusable until then in any case...

File: qa/rpc-tests/bigblocks.py (hunk 1 above)
Comment (sugruedes) [inline qa/rpc-tests/bigblocks.py:170]: Got that, thanks. I was looking at it from usual behaviour as opposed to specific test case behaviour.

File: qa/rpc-tests/bigblocks.py [hunk 3]
```diff
@@ -198,8 +214,22 @@ def run_test(self):
         # Time starts a little before fork activation time:
         set_node_times(self.nodes, FORK_TIME - 100)
 
-        # No supermajority
-        self.TestMineBig(False)
+        # No supermajority, vote2mb=1
```
Comment (dskloet) [inline qa/rpc-tests/bigblocks.py:217]: Use named parameters. Comments can be wrong and it's hard to spot.

File: qa/rpc-tests/bigblocks.py [hunk 4]
```diff
@@ -198,8 +214,22 @@ def run_test(self):
...
+        self.TestMineBig(False, True)
+
+        # Create a block after the expiration date. This will be rejected 
+        # by the other nodes for being more than 2 hours in the future,
```
Comment (dskloet) [inline qa/rpc-tests/bigblocks.py:221]: > This will be rejected  by the other nodes

Is that being asserted? Will that start failing once the expiry date has arrived?

**Comment by dskloet** [comment]
Maybe I missed it but if it's not already there, you should add a test that makes sure that if the fork happens, blocks >1MB will still be accepted after the fork bit expires. I'd hate to see the 1MB block limit return in 2018.

File: src/main.cpp (hunk 2 above)
Comment (jtoomim) [inline src/main.cpp:5321]: You're right, we can get rid of the cs_main lock. I wrote that assert because I was planning on doing MedianTimePast, and forgot to remove it later.

Yes, it can be moved to miner.cpp.

As for reserving (and using) the bit until 2018 regardless of the fork outcome, doing so would be a departure from the BIP9 specification. I do not see how the bit would be unusable after the fork completes if the fork completes successfully. Can you elaborate?

File: qa/rpc-tests/bigblocks.py (hunk 4 above)
Comment (jtoomim) [inline qa/rpc-tests/bigblocks.py:221]: > Is that being asserted? 

Yes, in a roundabout manner. After generating the block in the future, we do this:

```
    self.nodes[0].invalidateblock(b1hash)
    assert(self.sync_blocks(self.nodes))
```

That manually invalidates the block only on node0, and then syncs all 4 nodes. If the other 3 nodes accepted the block at first, they would never sync unless the block was manually invalidated. The fact that that manual invalidation is not necessary for nodes 1:4 indicates that they automatically invalidated the block.

File: qa/rpc-tests/bigblocks.py [hunk 5]
```diff
@@ -67,7 +68,10 @@ def setup_chain(self):
             for i in range(0,200):
                 miner = i%2
                 set_node_times(self.nodes, block_time)
-                self.nodes[miner].generate(1)
+                b1hash = self.nodes[miner].generate(1)[0]
+                b1 = self.nodes[miner].getblock(b1hash, True)
+                if miner % 2: assert(not (b1['version'] & FORK_BLOCK_BIT))
```
Comment (seusher) [inline qa/rpc-tests/bigblocks.py:73]: Obviously there isn't an easier way to get if the nodes are voting for 2mb or not, but since order matters, you may want to either put that comment above the node creation, or above the if block for assert statements. Its a bit fragile to depend on the order, but it would be a pain to extend the proxy to give access to the expose the vote via rpc.

**Comment by jonasschnelli** [comment]
Regardless of the dispute about blocksize and scaling:

A hard-fork that activates at 75% of hashpower is to dangerous and **I would recommend that Classic communicates the risks behind a such change better**. This will result in a sudden drop (max. of 25%) of bitcoin nodes security model with and potentially forms a fork (~25%) that will disrupt the market significant.

If one guy leaving bitcoin development can disrupt the market in a way we didn't expect, how deep could a HF with 75% activation threshold disrupt the market?

IMO if we want so see a long living Bitcoin, we cannot introduce hard forks that activates at 75%.

**Comment by jtoomim** [comment]
@jonasschnelli You're aware of the 4 week grace period, right? It won't be "a sudden drop" in hashrate or up-to-date nodes. Everyone will have time to upgrade.

Furthermore, you may note that there is a command-line option that allows miners to choose whether they want to vote for the fork or not. Those who vote for the fork will produce blocks with version 0x30000000, which get counted towards the supermajority. Those who do not vote for the fork will produce blocks with version 0x20000000, and will not count towards the supermajority. However, those who are producing version 0x20000000 blocks are still compatible with the fork, and will follow the longest chain. They will also switch to producing 0x30000000 blocks as soon as the supermajority threshold is exceeded.

**Comment by jonasschnelli** [comment]
> You're aware of the 4 week grace period, right? It won't be "a sudden drop" in hashrate or up-to-date nodes. Everyone will have time to upgrade.

No. I disagree.
The grace period is delusional.
Everybody if forced to follow during that time, or they stay on the "smaller" fork side. What if 25% don't want to and don't upgrade. It would result in a 25% lost of Bitcoin security.
What market disruptions would you expect from a such scenario?

I think a HF without overall consensus should not trigger at 75%, regardless of a grace period.
And I'm not arguing from a standpoint where I take position in the blocksize dispute.

**A HF with signs of non-full-consensus should never trigger at 75%**

**Comment by jtoomim** [comment]
@jonasschnelli It sounds like you think that up to 25% of Bitcoin users do not want to be part of the majority chain, and would prefer to use a < 1 MB branch with knowingly diminished security and very low usability. I find this hard to believe.

I also note that this particular PR has nothing to do with setting neither the grace period nor the activation threshold, it is merely test cases, an expiration date, and a bugfix.

**Comment by jonasschnelli** [comment]
> It sounds like you think that up to 25% of Bitcoin users do not want to be part of the majority chain, and would prefer to use a < 1 MB branch with knowingly diminished security and very low usability. I find this hard to believe.

I'm not saying it's not possible. Sure, you can create a major chain fork if you control 51% of the hashpower.
But: **What market disruptions would you expect in a such scenario?**
IMO: Bitcoin survived the last years because there was consensus. Users, the market, investors do trust in this model. A sudden break of the consensus model by a 75% HF, even if it 80%/20% (agreement on the HF) is totally unhealthy IMO.

Your free to go down that road. But I think it would be wise to inform your potential user base about the risks and consequences.

**Comment by jtoomim** [comment]
I think you're inventing an implausible what-if scenario.

One of the scenarios I want to avoid is the scenario where a single miner with 5% of the hashrate is subject to blackmail, bribery, or extortion in order to prevent the hard fork from proceeding. I'd like to be able to say that no daughters were kidnapped in the making of this fork.

Can we move this conversation off of github, please? Maybe put it on reddit, [here](https://www.reddit.com/r/Bitcoin_Classic/comments/40wnre/on_activation_thresholds/)?

**Comment by jonasschnelli** [comment]
@jtoomim: IMO It's about **consensus**. And in a consensus model, all should agree in the way to go. The whole economy sits in the same boat. Breaking that >95% consensus model – regardless of the technical possibilities – is a step backwards in development. Other models are already played through by banks and politics. I personally would really be disappointed to see bitcoin going into that (traditional non-modern) model.

But I don't want to interfere the technical work here. So I'm quiet now. :)

**I just wanted to make a point that a 75% activation is insane for the current market fragility and it would be a fundamental change in how changes where made in the past.**

**Comment by digitsu** [comment]
I think with all due respect Jonas, you are worrying about an impossible situation, (one where 25% don’t want to give up and join the majority chain, lose security for users, and lose money if they are miners due to slower block rates per kwh.) coupled with a mistaken idea of how Bitcoin comes to consensus.  We don’t come to consensus in “agreeing what fork to support -before it happen- with a super majority”.  We come to consensus through PoW (longest chain).  The situation you just described puts a meta layer of consensus (through threat of market volatility) on top of PoW.

If you want everyone to 95% agree -before- anyone decides to roll out any software, then I ask you, why bother even having PoW? Why not just use PoS?
Or any other byzantine fault tolerant system?

**Comment by jonasschnelli** [comment]
@digitsu:

> We come to consensus through PoW (longest chain)

Yes. But ideally the longest chain should reflect and respect all different parties in Bitcoin.
Why then going for 75% instead of 51%?
Why did we start accepting features of soft-forks at 75% but lock them irreversible at 95%?

If a miner does not respect all parties, or lets say major influence (>3%), he will indirectly negatively affect its own profit possibilities. IMO: miners should have a strong incentive to form consensus _NOT_ after PoW 51%, instead after a >95% consensus.

**Comment by gavinandresen** [comment]
NACK -- after wrestling with testing edge cases in unit tests for a day, I really don't like the added complexity in the voting rules; it violates the KISS rule.  Example of an edge case that requires more code and another unit test:

Vote succeeds at exactly the expiration date, but grace period is after the voting expiration date. If the voting bit is set during the grace period, does that mean a vote for a different soft fork?

Here's a KISS approach:
- The code will pay attention to, and continue to vote for, a 2MB increase until 1-Jan-2018.
- 2MB-bit-votes in blocks with timestamps after 1-Jan-2018 are counted as "unknown version".
- If the vote is not successful before 1-Jan-2018, the "This version of the software is obsolete" alert will be triggered, nagging the operator to upgrade.

I'll be trying to convince the BIP9 authors to simplify BIP9 so it just describes a way for forks to reserve version bits for a specific period of time.

File: qa/rpc-tests/bigblocks.py (hunk 4 above)
Comment (dskloet) [inline qa/rpc-tests/bigblocks.py:221]: So if it's asserted that the block will be rejected for being more than 2 hours in the future, doesn't that mean that the test will fail once the date arrives when it's no longer more than 2 hours in the future?

**Comment by dskloet** [comment]
@jonasschnelli @jtoomim why is this being discussed on this PR?

**Comment by jtoomim** [comment]
@gavinandresen:

> Vote succeeds at exactly the expiration date, but grace period is after the voting expiration date. If the voting bit is set during the grace period, does that mean a vote for a different soft fork?

The voting bit is not set during the grace period after the expiration date.

```
    if (nTime < sizeForkTime.load() &&
         nTime < consensusParams.SizeForkExpiration() &&
        (GetBoolArg("-vote2mb", DEFAULT_2MB_VOTE) || (pblocktree-ForkBitActivated(FORK_BIT_2MB) != uint256())))
```

Means:
- If 2 MB is not currently allowed, and
- If the fork voting period has not yet expired, and
- If either the user has chosen to vote for 2 MB or we are in the grace period, then
  = Mark the block for 2 MB.

I think that preventing votes from being cast after the expiration date makes things simpler for future developers. I think that preventing votes cast after the expiration date for a different fork that uses the same bit from being counted as in favor of the 2 MB fork also makes things simpler overall. Yes, it's a few more test cases that need to be written. I'm working on adding those too.

**Comment by ojanssens** [comment]
@jtoomim @gavinandresen We've decided to remove the command line option to vote for 2MB, so need need to continue working on this. 2MB will always be on when running this version.

**Comment by jtoomim** [comment]
@ojanssens The issue here is actually the expiration date, not the command-line voting.

By the way, when you said "need need to continue", I interpreted that to mean "no need to continue".
